### PR TITLE
Performance improvement, and associate API changes

### LINKED ISSUE #130: Change constructs access API to facilitate performance improvements
### Current situation

Access to named constructs is currently either by the "filter_by_*" `Constructs` methods, or by `Field` attributes:
```python
>>> f = cdm.example_field(0)
>>> f.dimension_coordinates
<Constructs: dimension_coordinate(3)>
>>> f.constructs.filter_by_type('dimension_coordinate')
<Constructs: dimension_coordinate(3)>
```
The reason for the "dimension_coordinates" alias being an attribute rather than method is that it was expected that it would often be followed by more options:
```python
>>> f.dimension_coordinates('time')
<Constructs: dimension_coordinate(1)>
>>> f.dimension_coordinates.filter_by_property(units='degrees_east')
<Constructs: dimension_coordinate(1)>
```

This is fine, but comes with the slight confusion that the result of the attribute is a callable `Constructs` instance, which obfuscates the help:
```python
>>> help(f.dimension_coordinates)
Help on Constructs in module cfdm.constructs object:

class Constructs(cfdm.mixin.container.Container, cfdm.core.constructs.Constructs)
 |  Constructs(auxiliary_coordinate=None, dimension_coordinate=None, domain_ancillary=None, field_ancillary=None, cell_measure=None, coordinate_reference=None, domain_axis=None, cell_method=None, source=None, copy=True, _use_data=True, _view=False, _ignore=())
...
```

### Performance

It turns out that the code for accessing constructs is very slow, and must be improved.

A key part of this improvement relies on being able to work with python dictionaries rather than dictionary-like `Constructs` objects. 

To facilitate this, whilst retaining the intuitive nature of the API (both in cfdm library code and in downstream applications) the least intrusive change is to make what were attributes properties: i.e. `f.dimension_coordinates` becomes `f.dimension_coordinates()`. This then allows keyword parameters that can change the behaviour when speed is an issue. 

### Code breaking

This change will break any code that uses bare construct access attributes:

```python
>>> # These won't work any more
>>> f.auxiliary_coordinates  
>>> f.coordinate_references
>>> f.coordinates  
>>> f.cell_measures
>>> f.dimension_coordinates
>>> f.domain_ancillaries
>>> f.domain_axes
>>> f.cell_methods
>>> f.field_ancillaries
```

```python
>>> # These will work as before
>>> f.auxiliary_coordinates()
>>> f.coordinate_references()
>>> f.coordinates()
>>> f.cell_measures()
>>> f.dimension_coordinates()
>>> f.domain_ancillaries()
>>> f.domain_axes()
>>> f.cell_methods()
>>> f.field_ancillaries()
```

```python
>>> # These will also work as before
>>> f.auxiliary_coordinates(x)
>>> f.coordinate_references(x)
>>> f.coordinates(x)
>>> f.cell_measures(x)
>>> f.dimension_coordinates(x)
>>> f.domain_ancillaries(x)
>>> f.domain_axes(x)
>>> f.cell_methods(x)
>>> f.field_ancillaries(x)
```

This is the only backwards incompatible change to the API. All other changes will not break existing code.

With the new API, reading a file is, in one reproducible test, **~10 times faster**:

```python
>>> import cfdm, timeit
>>> f = cfdm.example_field(1)
>>> cfdm.write(f, 'tmp.nc')
>>> cfdm.__version__
1.8.8.0
>>> sum(timeit.repeat("cfdm.read('tmp.nc')", globals=globals(), repeat=100, number=1))/100
0.25303671994010074
```

```python
>>> import cfdm, timeit
>>> cfdm.__version__
1.8.9.0b1
>>> sum(timeit.repeat("cfdm.read('tmp.nc')", globals=globals(), repeat=100, number=1))/100
0.026102047999302158
```
Note that much of the improvement comes from unrelated changes (such as removing unnecessary `__repr__` calls and unnecessary deep copies).

PR to follow ...

### LINKED ISSUE #132: Change constructs access API for consistency
### Current situation

The `construct` and `construct_key` methods take a single positional `identity` argument (https://github.com/NCAS-CMS/cfdm/blob/v1.8.8.0/cfdm/mixin/constructaccess.py#L301):
```python
    def construct(self, identity, default=ValueError()):
```

### Proposed change

The `construct` and `construct_key` methods take any number of positional arguments, as well as new keyword arguments:
```python
    def construct(self, *identity, default=ValueError(), **filter_kwargs):
```
The `**filter_kwargs` are passed to the new `Constructs.filter` method for further selection  flexibility.

This makes sense, as you can already match on multiple identities by providing a `re.Pattern` object, so it doesn't makes sense to restrict this to, say, `*["latitude", "longitude"]`

### Consequences

The only possibly bad effect that this could have is for the  `default` argument to be consumed by the `*identity` arguments if it is given as a positional parameter. For example, to Specify a default of `None`:

```python
>>> f.construct('altitude', None)  # ok at v1.8.8.0
>>> f.construct('altitude', None)  # NOT ok with proposed change
>>> f.construct('altitude', default=None)  # ok with proposed change and at v1.8.8.0
```

Regardless of this potential for code breaking, I think this change will be beneficial.

### LINKED ISSUE #137: Backwards incompatible change to Constructs.filter_by_axis API
The API changes described in #130 and #132 imply a change in `Constructs.filter_by_axis`, namely the existing:

```python 
    def filter_by_axis(mode=None, axes*):  # v1.8.8.0
```

becomes:

```python 
    def filter_by_axis(axes, axis_mode='and'):  # v1.8.9.0
```

The old `mode` and new `axis_mode` parameters behave the same.

### Potential for existing code to silently break 

The existing code

```python
>>> d = c.filter_by_axis('or', 'domainaxis1')  # v1.8.8.0
```
will behave differently with this change in v1.8.9.0, in that both `'or'` and `'domainaxis1''` will now count as `*axes` arguments , rather than the `'or'` being assigned to the old `mode` parameter.

The way to rewrite this code would be:

```python
>>> d = c.filter_by_axis('domainaxis1', axis_mode='or')  # v1.8.9.0
```

### PR DESCRIPTION
**Pull request by davidhassell**
Fixes #130 
Fixes #132 
Fixes #137

### CODE REVIEW

**Comment by davidhassell** [comment]
Hi @sadielbartholomew,

As discussed off-line, this is a rambling old PR (98 files changed), but most changes are either style, linting, or the implementation of the new API throughout the library.

When reviewing, you might search for `1.8.9.0` in the code, or I might recommend starting with (in some sort of logical order):

* `cfdm/core/constructs.py`
* `cfdm/constructs.py`
* `cfdm/mixin/container.py`
* `cfdm/mixin/*.py`
* `cfdm/mixin/fielddomain.py` [*]
* `cfdm/field.py`

[*] new name for `cfdm/constructaccess.py`, looking ahead to v1.9.0.0

You might get away with just that, but I'm sure that you could also treat these files as spring board to other things.

Thanks.

**Comment by davidhassell** [comment]
I notice that the GitHub action tests are failing with

```python
    from re import Pattern
ImportError: cannot import name 'Pattern'
```
on Python 3.6. 

.... Oh - I see that they were added in python 3.7: https://github.com/python/typeshed/issues/3447#issuecomment-550420561

That makes three good reasons to deprecate 3.6:

1. numpy 1.20 doesn't work with 3.6
2. dict is not ordered at 3.6
3. No re.Pattern object at 3.6

Number 2. already has a workaround and I'll put in a workaround for 3., for now

**Review by sadielbartholomew — APPROVED** [review]
Fantastic! Users should be very pleased with the performance gains. Great work.

I'm happy with the API changes proposed in the assocaited issues and satisfied that they are implemented successfully in this PR.

Regarding the code, I have made a few minor comments as provided in-line, and some suggestions as follows:

1. ...

For your reference, below I have outlined the results of some speed tests I conducted to gauge the performance gains. All tests go faster, though there wasn't nearly as much of a relative speed up as you have reported, especially in the latter case. Do you know why that might be?

### Observed performance improvements

#### Test suite

As a quick indication for the speed, as reported in the pytest message "Ran <N> tests in <M>s" rather than by using a `time` command), of running the full test suite in my local environment, across five runs on each branch I saw:

* current `master`: "Ran 171 tests in ..." between 30.2 and 32.3 s;
* this branch: "Ran 195 tests in ..." between 27.7 and 29.6 s.

So the suite is running faster despite having more tests to get through, which is great.

#### Snippet from (end of) https://github.com/NCAS-CMS/cfdm/issues/130#issue-846679007

Running that snippet I see:
* current `master`: ~0.021 (`0.020565794229769382`)
* this branch: ~0.017 (`0.016937515400059056`)
which is faster after this PR but not by nearly as much as suggested when you did the same test?

**Review by sadielbartholomew — COMMENTED** [review]
Just registering a few more minor comments to complete my review started above. Thanks.

File: cfdm/mixin/fielddomain.py
Comment (davidhassell) [reply cfdm/mixin/fielddomain.py:184]: Yes, they should not both be True - I'll add a note to the effect that if `key=True` then item is ignored. I think that `key=True` is more often used, so that's why I put it first.

File: cfdm/mixin/fielddomain.py
Comment (davidhassell) [reply cfdm/mixin/fielddomain.py:303]: Yes ... but we want to continue if `keys` has one element, too ...

File: cfdm/mixin/fielddomain.py
Comment (davidhassell) [reply cfdm/mixin/fielddomain.py:177]: This is resolved by putting this code into a helper function (https://github.com/NCAS-CMS/cfdm/pull/138#discussion_r628967012)

File: cfdm/mixin/container.py
Comment (davidhassell) [reply cfdm/mixin/container.py:206]: Fixed in next commit

File: cfdm/constructs.py
Comment (davidhassell) [reply cfdm/constructs.py:672]: Fixed in next commit

File: cfdm/constructs.py
Comment (davidhassell) [reply cfdm/constructs.py:565]: Indeed. Nonsense. I've changed to:

```
"""Parse arguments destined for a filter method.
```

File: cfdm/constructs.py
Comment (davidhassell) [reply cfdm/constructs.py:551]: Good spot. It has gone.

**Comment by davidhassell** [comment]
#### Snippet from (end of) #130 (comment)

I still get a factor of 10 speed up:

```python
In [1]: import cfdm, timeit

In [2]: f = cfdm.example_field(1)

In [3]: cfdm.write(f, 'tmp.nc')

In [4]: assert cfdm.__version__ == '1.8.9.0'

In [5]: sum(timeit.repeat("cfdm.read('tmp.nc')", globals=globals(), repeat=100, number=1))/100
Out[5]: 0.02655579654999656
```

```python
In [1]: import cfdm, timeit

In [2]: assert cfdm.__version__ == '1.8.8.0'

In [3]: sum(timeit.repeat("cfdm.read('tmp.nc')", globals=globals(), repeat=100, number=1))/100
Out[3]: 0.2563364723699988
```

???

**Comment by sadielbartholomew** [comment]
> I still get a factor of 10 speed up

:thinking: Ah, wait, sorry @davidhassell I was about to ask which `master` state you are using to compare to and but I see from the versioning outputs it is instead probably the tagged 1.8.8.0 release, in which case that could be the root of it because I was comparing against the current `master`. Still, I think my test is the fairer one for this PR, as testing against 1.8.8.0, though you acknowledge that it I think:

> Note that much of the improvement comes from unrelated changes (such as removing unnecessary __repr__ calls and unnecessary deep copies).

My bad, let me re-do the test comparing against 1.8.8.0, though it is noteworthy that we're only seeing factor ~1.25 speed up from this PR, at least in my environment. Still, that's good in itself!

File: cfdm/mixin/fielddomain.py
Comment (sadielbartholomew) [reply cfdm/mixin/fielddomain.py:342]: :laughing: Very good point, so sorry! Ignore this. That's me misinterpreting a simple operator and talking nonsense...

**Comment by sadielbartholomew** [comment]
OK when attempting comparison with the `v1.8.8.0` release branch, I keep seeing a seg fault :frowning_face:, even if I reduce the `repeat` parameter to 2 and also 1!:

```python
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cfdm, timeit
>>> f = cfdm.example_field(1)
>>> cfdm.write(f, 'tmp.nc')
>>> assert cfdm.__version__ == '1.8.8.0'
>>> sum(timeit.repeat("cfdm.read('tmp.nc')", globals=globals(), repeat=100, number=1))/100
Segmentation fault (core dumped)
```

which isn't useful but potentially illuminating in the seg faulting saga. I'm not sure I've seen one outside of the test suite before. Luckily I didn't get any such seg faults on the current `master` and/or this branch, so that could imply something has been fixed in a dependency else in our library...

Either way, because of that I can't do the comparison test with 1.8.8.0. Let me try an earlier version, one moment...

**Comment by sadielbartholomew** [comment]
Every release branch I try below `1.8.8.0` is seg faulting so let's assume there is some issue with my environment there and that I can't do the comparison test. I'm happy to take your factor of 10 comparison from 1.8.80 to `master` after this is merged as representative.

File: cfdm/mixin/container.py
Comment (sadielbartholomew) [reply cfdm/mixin/container.py]: Not quite, it looks like! We can always sort it later for the pre-release checks, though, if you prefer.

File: cfdm/mixin/fielddomain.py
Comment (sadielbartholomew) [reply cfdm/mixin/fielddomain.py]: Great, thanks for the clarification and the new note in the docstring which makes the context very clear.

**Comment by davidhassell** [comment]
Great - thanks @sadielbartholomew, as ever, for the careful review. I agree that my various PRs relating to this are not that clean (sorry!), but I think we've got there.

I'll merge when I've heard back from you on that typo (https://github.com/NCAS-CMS/cfdm/pull/138/commits/0a60a47df477a245ac57026d209e71e2bdc21e52), and done a final run of the unit tests (and checked that the upstream cf-python tests still pass).

I'm looking forward to reviewing the append mode PR later.

**Comment by sadielbartholomew** [comment]
All good, thanks @davidhassell. Please merge away!

**Pull request merged**
